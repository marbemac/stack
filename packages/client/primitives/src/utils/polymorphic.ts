/* eslint-disable @typescript-eslint/ban-types */
// Adapted from https://github.com/fabien-ml/kobalte/blob/8ab09d87f7ea68f9fb79d32458a06dd3ae62029d/packages/utils/src/polymorphic.tsx
// asProps design came out of this discussion -> https://github.com/orgs/kobaltedev/discussions/138#discussioncomment-5245471

import type { SetRequired } from '@marbemac/utils-types';
import type { Component, ComponentProps as BaseComponentProps, JSX } from 'solid-js';

/** All HTML and SVG elements. */
export type DOMElements = keyof JSX.IntrinsicElements;

/** Any HTML element or SolidJS component. */
export type ElementType<Props = any> = DOMElements | Component<Props>;

/**
 * Allows for extending a set of props (`Source`) by an overriding set of props (`Override`),
 * ensuring that any duplicates are overridden by the overriding set of props.
 */
export type OverrideProps<Source = {}, Override = {}> = Omit<Source, keyof Override> & Override;

/** The `as` prop type. */
export type As<Props = any> = ElementType<Props>;

/** Clean-up autocomplete and omit a bunch of props we don't use */
export type ComponentProps<T extends As> = Omit<BaseComponentProps<T>, keyof IgnoredHTMLAttributes<T>>;

export type IgnoredHTMLAttributes<T> = Pick<
  JSX.HTMLAttributes<T>,
  // Feel free to remove any items from this list if any end up being needed
  | '$ServerOnly'
  | 'about'
  | 'accessKey'
  | 'aria-atomic'
  | 'aria-flowto'
  | 'autocapitalize'
  | 'class'
  | 'classList'
  | 'color'
  | 'contenteditable'
  | 'contextmenu'
  | 'datatype'
  | 'dir'
  | 'exportParts'
  | 'exportparts'
  | 'inert'
  | 'inlist'
  | 'inputMode'
  | 'inputmode'
  | 'itemId'
  | 'itemProp'
  | 'itemRef'
  | 'itemScope'
  | 'itemType'
  | 'itemid'
  | 'itemprop'
  | 'itemref'
  | 'itemscope'
  | 'itemtype'
  | 'onabort'
  | 'onanimationend'
  | 'onanimationiteration'
  | 'onanimationstart'
  | 'onauxclick'
  | 'onbeforeinput'
  | 'onblur'
  | 'oncanplay'
  | 'oncanplaythrough'
  | 'onchange'
  | 'onclick'
  | 'oncontextmenu'
  | 'ondblclick'
  | 'ondrag'
  | 'ondragend'
  | 'ondragenter'
  | 'ondragleave'
  | 'ondragover'
  | 'ondragstart'
  | 'ondrop'
  | 'ondurationchange'
  | 'onemptied'
  | 'onended'
  | 'onerror'
  | 'onfocus'
  | 'ongotpointercapture'
  | 'oninput'
  | 'oninvalid'
  | 'onkeydown'
  | 'onkeypress'
  | 'onkeyup'
  | 'onload'
  | 'onloadeddata'
  | 'onloadedmetadata'
  | 'onloadstart'
  | 'onlostpointercapture'
  | 'onmousedown'
  | 'onmouseenter'
  | 'onmouseleave'
  | 'onmousemove'
  | 'onmouseout'
  | 'onmouseover'
  | 'onmouseup'
  | 'onpause'
  | 'onplay'
  | 'onplaying'
  | 'onpointercancel'
  | 'onpointerdown'
  | 'onpointerenter'
  | 'onpointerleave'
  | 'onpointermove'
  | 'onpointerout'
  | 'onpointerover'
  | 'onpointerup'
  | 'onprogress'
  | 'onratechange'
  | 'onreset'
  | 'onscroll'
  | 'onseeked'
  | 'onseeking'
  | 'onselect'
  | 'onstalled'
  | 'onsubmit'
  | 'onsuspend'
  | 'ontimeupdate'
  | 'ontouchcancel'
  | 'ontouchend'
  | 'ontouchmove'
  | 'ontouchstart'
  | 'ontransitionstart'
  | 'ontransitionend'
  | 'ontransitionrun'
  | 'ontransitioncancel'
  | 'onvolumechange'
  | 'onwaiting'
  | 'onwheel'
  | 'oncompositionend'
  | 'oncompositionstart'
  | 'oncompositionupdate'
  | 'oncopy'
  | 'oncut'
  | 'ondragexit'
  | 'onencrypted'
  | 'onfocusout'
  | 'onfocusin'
  | 'onpaste'
  | 'part'
  | 'prefix'
  | 'property'
  | 'resource'
  | 'slot'
  | 'tabindex'
  | 'typeof'
  | 'vocab'
>;

/** Props object that includes the `as` prop. */
export type PolymorphicProps<Type extends As = As, Props = {}> = OverrideProps<
  ComponentProps<Type>,
  // Props & { as?: Type }
  Props & AsProp<Type>
>;

interface AsProp<T extends As, P = ComponentProps<T>> {
  as?: T;
  asProps?: P;
}

type WithAs<T extends As> = SetRequired<AsProp<T>, 'as'>;

/** A component with the `as` prop. */
export type PolymorphicComponent<DefaultType extends As, Props = {}> = {
  <Type extends As>(props: PolymorphicProps<Type, Props> & WithAs<Type>): JSX.Element;
  (props: PolymorphicProps<DefaultType, Props>): JSX.Element;
};

/**
 * Create a component with the type cast to `PolymorphicComponent`.
 * You have to use `Dynamic` internally and pass the `as` prop to handle polymorphism correctly.
 */
export function createPolymorphicComponent<DefaultType extends As, Props = {}>(
  component: Component<PolymorphicProps<DefaultType, Props>>,
) {
  return component as unknown as PolymorphicComponent<DefaultType, Props>;
}
