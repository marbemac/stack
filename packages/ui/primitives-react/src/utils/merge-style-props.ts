/**
 * Merge default and component props into a single props object.
 * @param defaultProps The default props, will be overridden by component props.
 * @param props The component `props` object.
 * @example
 * // mergedProps = defaultProps <== props
 */
export function mergeStyleProps<T extends Record<string, any>>(defaultProps: Partial<T>, props: T): T {
  return Object.assign({}, defaultProps, props);
}
