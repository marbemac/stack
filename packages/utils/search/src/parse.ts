import type { ILexingResult, IRecognitionException, Lexer } from 'chevrotain';

import { SearchLexer } from './grammar/lexer.ts';
import { SearchParser } from './grammar/parser.ts';
import type { SearchString } from './types.ts';

let lexerSingleton: Lexer;
let parserSingleton: SearchParser;

export interface ParseQueryOpts<T extends InputType> {
  input: SearchString;
  inputType: T;
  lexer?: Lexer;
  parser?: SearchParser;
}

type InputType = 'searchQuery' | 'selectClause' | 'selectExpression' | 'fromClause' | 'whereClause' | 'whereExpression';

export const parseSearch = <T extends InputType>({
  input,
  inputType,
  ...rest
}: ParseQueryOpts<T>): {
  cst: ReturnType<SearchParser[T]>;
  errors: IRecognitionException[];
  lexResult: ILexingResult;
} => {
  let lexer = rest.lexer || lexerSingleton;
  if (!lexer) {
    lexer = lexerSingleton = new SearchLexer();
  }

  const lexResult = lexer.tokenize(input);

  let parser = rest.parser || parserSingleton;
  if (!parser) {
    parser = parserSingleton = new SearchParser();
  }

  // "input" is a setter which will reset the parser's state.
  parser.input = lexResult.tokens;

  const cst = parser[inputType]() as ReturnType<SearchParser[T]>;

  return { cst, errors: parser.errors, lexResult };
};
