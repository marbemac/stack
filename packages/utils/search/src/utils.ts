import {
  type FunctionAstNode,
  type IsNegatableNode,
  type IsSortableNode,
  type QualifierAstNode,
  type SearchNode,
} from './visitor.ts';

export function stringifySearchAstNode(token?: SearchNode) {
  if (!token) return '';

  switch (token.type) {
    case 'search_query':
      return [
        stringifySearchAstNode(token.fromClause),
        stringifySearchAstNode(token.selectClause),
        stringifySearchAstNode(token.whereClause),
      ]
        .filter(Boolean)
        .join('\n');
    case 'from_clause':
      return `FROM ${token.table}`;
    case 'select_clause':
      return token.expr.columns.length ? `SELECT ` : '';
    case 'select_expr':
      return token.columns.length ? token.columns.map(stringifySearchAstNode).join(' ') : '';
    case 'where_clause':
      return token.expr.conditions.length ? `WHERE ` : '';
    case 'where_expr':
      return token.conditions.length ? token.conditions.map(stringifySearchAstNode).join(' ') : '';
    case 'qualifier':
      return handleNegation(token, stringifyQualifier(token));
    case 'function':
      return handleSortDir(token, handleNegation(token, stringifyFunction(token)));
    case 'function_arg':
      return token.vals.map(stringifySearchAstNode).join(' ');
    case 'qualifier_key':
      return token.value;
    case 'bracket_list':
      return `[${token.values.map(stringifySearchAstNode).join(', ')}]`;
    case 'relative_date':
      return `${token.sign}${token.value}${token.unit}`;
    case 'text':
      return handleSortDir(token, token.quoted ? `"${token.value}"` : token.value);
    case 'number':
      return token.value;
    case 'boolean':
      return token.value;
  }
}

function stringifyQualifier(token: QualifierAstNode) {
  let stringifiedToken = stringifySearchAstNode(token.lhs);

  if (token.rhs) {
    stringifiedToken += ':';
    stringifiedToken += token.op || '';
    stringifiedToken += stringifySearchAstNode(token.rhs);
  }

  return stringifiedToken;
}

function stringifyFunction(token: FunctionAstNode) {
  let stringifiedToken = `${token.name}(`;

  if (token.args) {
    stringifiedToken += token.args.map(stringifySearchAstNode).join(', ');
  }

  stringifiedToken += ')';

  if (token.rhs) {
    stringifiedToken += ':';
    stringifiedToken += token.op || '';
    stringifiedToken += stringifySearchAstNode(token.rhs);
  }

  return stringifiedToken;
}

function handleSortDir(token: IsSortableNode, stringifiedToken: string) {
  if (!token.sort) return stringifiedToken;
  return `${token.sort === 'asc' ? '+' : '-'}${stringifiedToken}`;
}

function handleNegation(token: IsNegatableNode, stringifiedToken: string) {
  return token.negated ? `!${stringifiedToken}` : stringifiedToken;
}
