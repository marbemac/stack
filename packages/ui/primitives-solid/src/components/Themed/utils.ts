import { isFunction } from '@kobalte/utils';
import { mergeProps } from 'solid-js';

import type { ComponentsConfig } from './components-config.ts';
import { useTheme } from './theme-context.ts';

function useComponentTheme<T extends keyof ComponentsConfig>(component: T) {
  return useTheme()?.components?.[component] ?? undefined;
}

/**
 * Resolve the `slotClasses` provided in the component theme configuration.
 */
export function useThemeClasses<Slots extends string>(component: keyof ComponentsConfig, props: any) {
  const classes = useComponentTheme(component)?.slotClasses ?? {};

  return (isFunction(classes) ? classes(props) : classes) as Partial<Record<Slots, string>>;
}

/**
 * Merge default, theme and component props into a single props object.
 * @param component The name of the component to look for in the theme.
 * @param defaultProps The default props, will be overridden by theme and component props.
 * @param props The component `props` object.
 * @example
 * // mergedProps = defaultProps <== themeProps <== props
 */
export function mergeThemeProps<T extends Record<string, any>>(
  component: keyof ComponentsConfig,
  defaultProps: Partial<T>,
  props: T,
): T {
  const themeProps = () => useComponentTheme(component)?.defaultProps ?? {};

  return mergeProps(defaultProps, themeProps, props);
}
