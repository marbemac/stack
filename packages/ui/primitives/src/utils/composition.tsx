/**
 * Credits to @ariakit for most of this code, and the "render" composition pattern.
 */

import { getRefProperty, mergeProps } from '@ariakit/react-core/utils/misc';
import { cloneElement, isValidElement } from 'react';

import { mergeRefs } from './merge-refs.ts';

/**
 * Render prop type.
 * @template P Props
 * @example
 * const children: RenderProp = (props) => <div {...props} />;
 */
export type RenderProp<P = React.HTMLAttributes<any> & React.RefAttributes<any>> = (props: P) => React.ReactNode;

/**
 * The `wrapElement` prop.
 */
export type WrapElement = (element: React.ReactElement) => React.ReactElement;

/**
 * Custom props including the `render` prop.
 */
export interface Options {
  wrapElement?: WrapElement;
  /**
   * Allows the component to be rendered as a different HTML element or React
   * component. The value can be a React element or a function that takes in the
   * original component props and gives back a React element with the props
   * merged.
   *
   * Check out the [Composition](https://ariakit.org/guide/composition) guide
   * for more details.
   */
  render?: RenderProp | React.ReactElement | keyof React.JSX.IntrinsicElements;
}

/**
 * Any object.
 */
export type AnyObject = Record<string, any>;

/**
 * Empty object.
 */
export type EmptyObject = Record<keyof any, never>;

/**
 * Any function.
 */
export type AnyFunction = (...args: any) => any;

/**
 * HTML props based on the element type, excluding custom props.
 * @template T The element type.
 * @template P Custom props.
 * @example
 * type ButtonHTMLProps = HTMLProps<"button", { custom?: boolean }>;
 */
export type HTMLProps<T extends React.ElementType, P extends AnyObject = EmptyObject> = Omit<
  React.ComponentPropsWithRef<T>,
  keyof P
> &
  Record<`data-${string}`, unknown>;

/**
 * Props based on the element type, including custom props.
 * @template T The element type.
 * @template P Custom props.
 */
export type Props<T extends React.ElementType, P extends AnyObject = EmptyObject> = P & HTMLProps<T, P>;

/**
 * Creates a React element that supports the `render` and `wrapElement` props.
 */
export function createElement(Type: React.ElementType, props: Props<React.ElementType, Options>) {
  const { wrapElement, render, ...rest } = props;
  const mergedRef = mergeRefs(props['ref'], getRefProperty(render));

  let element: React.ReactElement;

  if (isValidElement<any>(render)) {
    const renderProps = { ...render.props, ref: mergedRef };
    element = cloneElement(render, mergeProps(rest, renderProps));
  } else if (typeof render === 'string') {
    element = createElement(render, rest);
  } else if (render) {
    element = render(rest) as React.ReactElement;
  } else {
    element = <Type {...rest} />;
  }

  if (wrapElement) {
    return wrapElement(element);
  }

  return element;
}
