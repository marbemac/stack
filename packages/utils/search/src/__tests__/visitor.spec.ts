import { describe, expect, it } from 'vitest';

import { parseQuery } from '../parse.ts';
import type { SearchString } from '../types.ts';
import { treeTransformer } from '../utils.ts';
import { createSearchVisitor } from '../visitor.ts';
import { caseGroups } from './cases.ts';

const prettyPrintParseResult = (input: string) => {
  const { cst } = parseQuery({ input: input as SearchString, inputType: 'whereExpression' });

  const visitor = createSearchVisitor();
  const ast = visitor.whereExpression(cst.children);

  /**
   * Just cleans up the snapshots a little bit for easier review
   */
  return treeTransformer({
    tree: ast,
    transform: token => {
      // @ts-expect-error remove `quoted` property if false, since that's the default
      if (token.quoted === false) delete token.quoted;

      // @ts-expect-error remove `negated` property if false, since that's the default
      if (token.negated === false) delete token.negated;

      return token;
    },
  });
};

describe.each(Object.entries(caseGroups))('%s', (_, cases) => {
  it.each(cases)('%s', (_, input) => {
    expect({
      input,
      result: prettyPrintParseResult(input),
    }).toMatchSnapshot();
  });
});

it('debug', { skip: true }, () => {
  // const input = 'one: two():boo';
  // const input = 'one:two "foo bar" three:four';
  const input = 'count(projects, plan:foo jam:2)';

  const lexResult = { tokens: [] };
  // const { lexResult } = parseQuery({ input: input as SearchString, inputType: 'whereExpression' });

  expect(lexResult.tokens).toMatchInlineSnapshot(`[]`);

  expect(prettyPrintParseResult(input)).toMatchInlineSnapshot(`
    [
      {
        "lhs": {
          "args": [
            [
              {
                "quoted": false,
                "type": "atomicFilterVal",
                "value": "projects",
              },
            ],
            [
              {
                "lhs": {
                  "type": "filterKey",
                  "value": "plan",
                },
                "negated": false,
                "op": "=",
                "rhs": {
                  "quoted": false,
                  "type": "atomicFilterVal",
                  "value": "foo",
                },
                "type": "filter",
              },
              {
                "lhs": {
                  "type": "filterKey",
                  "value": "jam",
                },
                "negated": false,
                "op": "=",
                "rhs": {
                  "quoted": false,
                  "type": "atomicFilterVal",
                  "value": "2",
                },
                "type": "filter",
              },
            ],
          ],
          "name": "count",
          "type": "function",
        },
        "negated": false,
        "op": "=",
        "rhs": undefined,
        "type": "filter",
      },
    ]
  `);
});
