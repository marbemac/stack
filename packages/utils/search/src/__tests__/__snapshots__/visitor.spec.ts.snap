// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`freeform > mixed throughout 1`] = `
{
  "input": "freeform1 age:5 freeform2 search",
  "result": [
    {
      "type": "textVal",
      "value": "freeform1",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "age",
      },
      "op": "",
      "rhs": {
        "type": "numberVal",
        "value": "5",
      },
      "type": "filter",
    },
    {
      "type": "textVal",
      "value": "freeform2",
    },
    {
      "type": "textVal",
      "value": "search",
    },
  ],
}
`;

exports[`freeform > quoted 1`] = `
{
  "input": ""a quoted:freeform" age:5",
  "result": [
    {
      "quoted": true,
      "type": "textVal",
      "value": "a quoted:freeform",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "age",
      },
      "op": "",
      "rhs": {
        "type": "numberVal",
        "value": "5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`freeform > treats as a string 1`] = `
{
  "input": "freeform",
  "result": [
    {
      "type": "textVal",
      "value": "freeform",
    },
  ],
}
`;

exports[`functions > no args 1`] = `
{
  "input": "full_name()",
  "result": [
    {
      "lhs": {
        "args": [],
        "name": "full_name",
        "type": "function",
      },
      "op": "",
      "rhs": undefined,
      "type": "filter",
    },
  ],
}
`;

exports[`functions > no args with space 1`] = `
{
  "input": "full_name(  )",
  "result": [
    {
      "lhs": {
        "args": [],
        "name": "full_name",
        "type": "function",
      },
      "op": "",
      "rhs": undefined,
      "type": "filter",
    },
  ],
}
`;

exports[`functions > single string arg 1`] = `
{
  "input": "count(events)",
  "result": [
    {
      "lhs": {
        "args": [
          [
            {
              "type": "textVal",
              "value": "events",
            },
          ],
        ],
        "name": "count",
        "type": "function",
      },
      "op": "",
      "rhs": undefined,
      "type": "filter",
    },
  ],
}
`;

exports[`functions > with filter argument 1`] = `
{
  "input": "count(projects, plan:free !has:email):>5",
  "result": [
    {
      "lhs": {
        "args": [
          [
            {
              "type": "textVal",
              "value": "projects",
            },
          ],
          [
            {
              "lhs": {
                "type": "filterKey",
                "value": "plan",
              },
              "op": "",
              "rhs": {
                "type": "textVal",
                "value": "free",
              },
              "type": "filter",
            },
            {
              "lhs": {
                "type": "filterKey",
                "value": "has",
              },
              "negated": true,
              "op": "",
              "rhs": {
                "type": "textVal",
                "value": "email",
              },
              "type": "filter",
            },
          ],
        ],
        "name": "count",
        "type": "function",
      },
      "op": ">",
      "rhs": {
        "type": "numberVal",
        "value": "5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`functions > with quoted argument 1`] = `
{
  "input": "count(projects, "free search"):>0.5",
  "result": [
    {
      "lhs": {
        "args": [
          [
            {
              "type": "textVal",
              "value": "projects",
            },
          ],
          [
            {
              "quoted": true,
              "type": "textVal",
              "value": "free search",
            },
          ],
        ],
        "name": "count",
        "type": "function",
      },
      "op": ">",
      "rhs": {
        "type": "numberVal",
        "value": "0.5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`functions > with rhs 1`] = `
{
  "input": "divide(projects, max_projects):>0.5",
  "result": [
    {
      "lhs": {
        "args": [
          [
            {
              "type": "textVal",
              "value": "projects",
            },
          ],
          [
            {
              "type": "textVal",
              "value": "max_projects",
            },
          ],
        ],
        "name": "divide",
        "type": "function",
      },
      "op": ">",
      "rhs": {
        "type": "numberVal",
        "value": "0.5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`in filters > negation 1`] = `
{
  "input": "!user:[jane, john]",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "user",
      },
      "negated": true,
      "op": "",
      "rhs": {
        "type": "filterIn",
        "values": [
          {
            "type": "textVal",
            "value": "jane",
          },
          {
            "type": "textVal",
            "value": "john",
          },
        ],
      },
      "type": "filter",
    },
  ],
}
`;

exports[`in filters > numbers 1`] = `
{
  "input": "release:[12.0]",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "release",
      },
      "op": "",
      "rhs": {
        "type": "filterIn",
        "values": [
          {
            "type": "numberVal",
            "value": "12.0",
          },
        ],
      },
      "type": "filter",
    },
  ],
}
`;

exports[`in filters > quoted 1`] = `
{
  "input": "release:[jane, "john doe"]",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "release",
      },
      "op": "",
      "rhs": {
        "type": "filterIn",
        "values": [
          {
            "type": "textVal",
            "value": "jane",
          },
          {
            "quoted": true,
            "type": "textVal",
            "value": "john doe",
          },
        ],
      },
      "type": "filter",
    },
  ],
}
`;

exports[`in filters > works 1`] = `
{
  "input": "user:[jane, john]",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "user",
      },
      "op": "",
      "rhs": {
        "type": "filterIn",
        "values": [
          {
            "type": "textVal",
            "value": "jane",
          },
          {
            "type": "textVal",
            "value": "john",
          },
        ],
      },
      "type": "filter",
    },
  ],
}
`;

exports[`operators > < 1`] = `
{
  "input": "num_members:<5.0",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "num_members",
      },
      "op": "<",
      "rhs": {
        "type": "numberVal",
        "value": "5.0",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`operators > <= 1`] = `
{
  "input": "num_members:<=5.0",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "num_members",
      },
      "op": "<=",
      "rhs": {
        "type": "numberVal",
        "value": "5.0",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`operators > = 1`] = `
{
  "input": "num_members:=5",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "num_members",
      },
      "op": "=",
      "rhs": {
        "type": "numberVal",
        "value": "5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`operators > > 1`] = `
{
  "input": "num_members:>5.0",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "num_members",
      },
      "op": ">",
      "rhs": {
        "type": "numberVal",
        "value": "5.0",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`operators > >= 1`] = `
{
  "input": "num_members:>=5",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "num_members",
      },
      "op": ">=",
      "rhs": {
        "type": "numberVal",
        "value": "5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`relative date > + 1`] = `
{
  "input": "renews_at:+30d",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "renews_at",
      },
      "op": "",
      "rhs": {
        "sign": "+",
        "type": "relativeDateVal",
        "unit": "d",
        "value": "30",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`relative date > - 1`] = `
{
  "input": "last_seen:-1d",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "last_seen",
      },
      "op": "",
      "rhs": {
        "sign": "-",
        "type": "relativeDateVal",
        "unit": "d",
        "value": "1",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`relative date > decimal 1`] = `
{
  "input": "renews_at:+1.2s",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "renews_at",
      },
      "op": "",
      "rhs": {
        "sign": "+",
        "type": "relativeDateVal",
        "unit": "s",
        "value": "1.2",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`relative date > with other filters 1`] = `
{
  "input": "plan:free renews_at:+4w "free form"",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "plan",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "free",
      },
      "type": "filter",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "renews_at",
      },
      "op": "",
      "rhs": {
        "sign": "+",
        "type": "relativeDateVal",
        "unit": "w",
        "value": "4",
      },
      "type": "filter",
    },
    {
      "quoted": true,
      "type": "textVal",
      "value": "free form",
    },
  ],
}
`;

exports[`text filters > .-_ in keys 1`] = `
{
  "input": "payments.plan-type_is:free",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "payments.plan-type_is",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "free",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > @.-_ in values 1`] = `
{
  "input": "email:jane@example.com type:public-company_x",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "email",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "jane@example.com",
      },
      "type": "filter",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "type",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "public-company_x",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > multiple 1`] = `
{
  "input": "plan:free age:5",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "plan",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "free",
      },
      "type": "filter",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "age",
      },
      "op": "",
      "rhs": {
        "type": "numberVal",
        "value": "5",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > negation 1`] = `
{
  "input": "!plan:free",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "plan",
      },
      "negated": true,
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "free",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > negation in quoted 1`] = `
{
  "input": "!user:"john doe"",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "user",
      },
      "negated": true,
      "op": "",
      "rhs": {
        "quoted": true,
        "type": "textVal",
        "value": "john doe",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > partial filter still parses 1`] = `
{
  "input": "age:> plan:free user:",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "age",
      },
      "op": ">",
      "rhs": undefined,
      "type": "filter",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "plan",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "free",
      },
      "type": "filter",
    },
    {
      "lhs": {
        "type": "filterKey",
        "value": "user",
      },
      "op": "",
      "rhs": undefined,
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > quoted value 1`] = `
{
  "input": "user:"john doe"",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "user",
      },
      "op": "",
      "rhs": {
        "quoted": true,
        "type": "textVal",
        "value": "john doe",
      },
      "type": "filter",
    },
  ],
}
`;

exports[`text filters > works 1`] = `
{
  "input": "plan:free",
  "result": [
    {
      "lhs": {
        "type": "filterKey",
        "value": "plan",
      },
      "op": "",
      "rhs": {
        "type": "textVal",
        "value": "free",
      },
      "type": "filter",
    },
  ],
}
`;
